File: Tutorials/Codes/Advanced_Robotics_Course/Chapter18/Lesson1/Chapter18_Lesson1.cpp
#include <vector>
#include <cmath>
#include <cstddef>

double wrapAngle(double angle)
{
    const double pi = 3.141592653589793;
    double a = std::fmod(angle + pi, 2.0 * pi);
    if (a < 0.0) {
        a += 2.0 * pi;
    }
    return a - pi;
}

// weights.size() == q1.size() == q2.size()
// revolute[i] is true if joint i is revolute.
double configDistance(const std::vector<double> &q1,
                      const std::vector<double> &q2,
                      const std::vector<double> &weights,
                      const std::vector<bool> &revolute)
{
    const std::size_t n = q1.size();
    double sum = 0.0;

    for (std::size_t i = 0; i < n; ++i) {
        double diff = q2[i] - q1[i];
        if (revolute[i]) {
            diff = wrapAngle(diff);
        }
        sum += weights[i] * diff * diff;
    }
    return std::sqrt(sum);
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter18/Lesson1/Chapter18_Lesson1.java
public final class ConfigMetric {
    private final double[] weights;
    private final boolean[] revolute;

    public ConfigMetric(double[] weights, boolean[] revolute) {
        if (weights.length != revolute.length) {
            throw new IllegalArgumentException("Dimension mismatch");
        }
        this.weights = weights.clone();
        this.revolute = revolute.clone();
    }

    private static double wrapAngle(double angle) {
        double pi = Math.PI;
        double a = (angle + pi) % (2.0 * pi);
        if (a < 0.0) {
            a += 2.0 * pi;
        }
        return a - pi;
    }

    public double distance(double[] q1, double[] q2) {
        if (q1.length != q2.length || q1.length != weights.length) {
            throw new IllegalArgumentException("Dimension mismatch");
        }
        double sum = 0.0;
        for (int i = 0; i < q1.length; ++i) {
            double diff = q2[i] - q1[i];
            if (revolute[i]) {
                diff = wrapAngle(diff);
            }
            sum += weights[i] * diff * diff;
        }
        return Math.sqrt(sum);
    }
}
      
